### DISCUSSION TITLE
Missing ts_rank for text search

### DISCUSSION THREAD

**Discussion by prabhatk579** [sqlalchemy/sqlalchemy#11177] [Usage Questions]
Hi

I'm trying to rank the documents that I retrieve from the pgvector using ts_rank, but I'm getting an error:
```bash
sqlalchemy.exc.ProgrammingError: (psycopg2.errors.UndefinedFunction) function ts_rank_cd(text, tsquery) does not exist
LINE 1: ...ted_at, vectors.updated_at AS vectors_updated_at, ts_rank_cd...
                                                             ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
```
Here is the code snippet:

```py
text_search_config: str = "english"

class REGCONFIG(UserDefinedType):
            def get_col_spec(self) -> str:
                return "regconfig"

if query_str is None:
    msg = "query_str must be specified for a sparse vector query."
    raise ValueError(msg)

ts_query = func.plainto_tsquery(
    type_coerce(cls.text_search_config, REGCONFIG),
    query_str,
)
dialect_cast = cast(Vectors.text, Text)  # Adjust for your dialect (e.g., String)
rank = func.ts_rank_cd(dialect_cast, ts_query).label("rank")

return (
    session.query(
        cls,
        rank,
    )
    .filter(<some filters>)
    .order_by(asc("rank"))
    .limit(limit)
    .all()
)
```
How can I rank the retrieved documents or provide them with a score post-retrieval?

I've also gone through:
- https://docs.sqlalchemy.org/en/20/core/sqlelement.html#sqlalchemy.sql.expression.func
- https://www.postgresql.org/docs/current/textsearch-intro.html

**Comment by CaselIT** [comment]
Hi,

You seem to be missing a cast or similar. The docs of postgresql show this for the ts_rank_cd:
`ts_rank_cd([ weights float4[], ] vector tsvector, query tsquery [, normalization integer ]) returns float4`
so you likely need to cast `dialect_cast` to `tsvector`, not text
https://www.postgresql.org/docs/current/textsearch-controls.html#TEXTSEARCH-RANKING

**Reply by CaselIT** [reply]
also on v2 you can import REGCONFIG from the postgresql dialect
